code/obj/item/device/pda2/pda2.dm

```diff
@@ -838,6 +856,42 @@
 		src.underlays += src.ID_image
 		src.updateSelfDialog()
 		user.UpdateName()
+
+	proc/eject_pen(var/mob/user as mob)
+		if (src.pen)
+			if (istype(user))
+				user.put_in_hand_or_drop(src.pen)
+			else
+				var/turf/T = get_turf(src)
+				src.pen.set_loc(T)
+			src.pen = null
+			src.UpdateOverlays(null, "pen")
+			return
+
+	proc/insert_pen(obj/item/insertedPen, mob/user)
+		if (!istype(insertedPen))
+			return
+		if (user)
+			user.u_equip(insertedPen)
+			src.pen = insertedPen
+			if(istype(insertedPen, /obj/item/clothing/mask/cigarette))
+				src.UpdateOverlays(image(src.icon, "cig"), "pen")
+			else if(istype(insertedPen, /obj/item/pen/crayon))
+				var/image/pen_overlay = image(src.icon, "crayon")
+				pen_overlay.color = insertedPen.color
+				src.UpdateOverlays(pen_overlay, "pen")
+			else if(istype(insertedPen, /obj/item/pen/pencil))
+				src.UpdateOverlays(image(src.icon, "pencil"), "pen")
+			else
+				src.UpdateOverlays(image(src.icon, "pen"), "pen")
+			var/original_icon_state = src.icon_state
+			animate(src, time=0, icon_state="")
+			animate(time=2, icon_state=original_icon_state)
+			animate(time=2, transform=matrix(null, 0, -1, MATRIX_TRANSLATE))
+			animate(time=3, transform=null)
+		insertedPen.set_loc(src)
```

Comment:
Minor code-reading thing: I'd move the loc-setting to be right close to the u_equip() call above as it's related; I initially thought you missed it until I kept reading.